{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": 1,
      "metadata": {
        "id": "UAkHrdU0XOck"
      },
      "outputs": [],
      "source": [
        "import numpy as np\n",
        "import pandas as pd\n",
        "from sklearn.model_selection import train_test_split\n",
        "from sklearn.preprocessing import MinMaxScaler\n",
        "import json\n",
        "\n",
        "file_path = \"/content/V2_data.xlsx\""
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "sheet_name = \"Results\"\n",
        "\n",
        "df = pd.read_excel(\n",
        "    file_path,\n",
        "    sheet_name=sheet_name,\n",
        "    header=7,            # headers start on row 8\n",
        "    engine=\"openpyxl\"\n",
        ")\n",
        "\n",
        "print(\"Raw shape:\", df.shape)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Cvl5p8mrYPgP",
        "outputId": "3cdf6d90-e96e-426c-9371-559921ec0276"
      },
      "execution_count": 8,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Raw shape: (4329, 13)\n"
          ]
        },
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "/usr/local/lib/python3.12/dist-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed\n",
            "  warn(msg)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "df.columns = (\n",
        "    df.columns\n",
        "      .astype(str)\n",
        "      .str.strip()\n",
        "      .str.replace(\"\\n\", \"\", regex=False)\n",
        "      .str.replace(\"\\r\", \"\", regex=False)\n",
        ")\n",
        "\n",
        "print(\"Cleaned columns:\")\n",
        "for col in df.columns:\n",
        "    print(repr(col))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "RkT6d5uGpIKb",
        "outputId": "0bb56e81-5bd3-41d6-c10e-665ddeb5b282"
      },
      "execution_count": 9,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cleaned columns:\n",
            "'Construct Barcode'\n",
            "'Extended Spacer(NNNN[20nt]NGGNNN)'\n",
            "'Strand'\n",
            "'Transcript'\n",
            "'Gene Symbol'\n",
            "'Amino Acid Cut position'\n",
            "'Percent Peptide'\n",
            "'Annotation'\n",
            "'sgRNA Score'\n",
            "'Low Flag'\n",
            "'AZD_200nM'\n",
            "'6TG_2ug/mL'\n",
            "'PLX_2uM'\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "required_cols = [\n",
        "    \"Construct Barcode\",\n",
        "    \"Extended Spacer(NNNN[20nt]NGGNNN)\",\n",
        "    \"Gene Symbol\",\n",
        "    \"Amino Acid Cut position\",\n",
        "    \"Percent Peptide\",\n",
        "]"
      ],
      "metadata": {
        "id": "PbZ6CZ4Fq1py"
      },
      "execution_count": 10,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# VERIFICATION STEP\n",
        "missing_cols = [c for c in required_cols if c not in df.columns]\n",
        "\n",
        "if missing_cols:\n",
        "    raise ValueError(\n",
        "        f\"Missing required columns: {missing_cols}\\n\"\n",
        "        f\"Available columns: {df.columns.tolist()}\"\n",
        "    )\n",
        "\n",
        "print(\"All required columns found.\")\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "n6B5RWEvq92o",
        "outputId": "f4cedf1c-57b1-4896-c27c-8477c5254311"
      },
      "execution_count": 11,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "All required columns found.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# drop rows with missing values\n",
        "before = df.shape[0]\n",
        "\n",
        "df = df.dropna(subset=required_cols).reset_index(drop=True)\n",
        "\n",
        "after = df.shape[0]\n",
        "print(f\"Rows before dropna: {before}\")\n",
        "print(f\"Rows after dropna : {after}\")"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "rMI7AZInrFN0",
        "outputId": "9fdf2b15-4f6b-43c6-c3a6-9f03e5ebf25e"
      },
      "execution_count": 12,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Rows before dropna: 4329\n",
            "Rows after dropna : 4328\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# one-hot encode 20-mer sequence\n",
        "BASE2IDX = {\"A\": 0, \"C\": 1, \"G\": 2, \"T\": 3}\n",
        "\n",
        "def one_hot_20mer(seq, length = 20):\n",
        "  seq = str(seq).upper().strip()\n",
        "  if len(seq) != length:\n",
        "    raise ValueError(f\"Sequence length {len(seq)} != {length} for: {seq}\")\n",
        "\n",
        "  arr = np.zeros((4, length), dtype = np.float32)\n",
        "  for i, base in enumerate(seq):\n",
        "    idx = BASE2IDX.get(base, None)\n",
        "    if idx is None:\n",
        "      raise ValueError(f\"Invalid base '{base}' in sequence: {seq}\")\n",
        "    arr[idx, i] = 1.0\n",
        "\n",
        "  return arr\n",
        "\n",
        "\n"
      ],
      "metadata": {
        "id": "i1RCQzG9rMOM"
      },
      "execution_count": 13,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "seq_list = []\n",
        "for s in df[\"Construct Barcode\"]:\n",
        "  seq_list.append(one_hot_20mer(s))\n",
        "\n",
        "X_seq = np.stack(seq_list, axis = 0)\n",
        "print(\"X_seq shape:\", X_seq.shape)\n",
        "\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ogBQRznvsHgQ",
        "outputId": "567e8b78-6d98-4b63-9413-6015f0c3baeb"
      },
      "execution_count": 14,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "X_seq shape: (4328, 4, 20)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Extract and encode PAM from Extended Spacer\n",
        "\n",
        "def extracct_pam(ext_seq):\n",
        "  \"\"\"\n",
        "  Default: positions 24:27 (NNNN[20nt]NGGNNN). If string is shorter/odd, use last 3 bases as fallback\n",
        "  \"\"\"\n",
        "\n",
        "  s = str(ext_seq).upper().replace(\" \",\"\")\n",
        "  if len(s) >= 27:\n",
        "    pam = s[24:27]\n",
        "  else:\n",
        "    pam = s[-3:]\n",
        "  return pam\n",
        "\n",
        "df[\"PAM\"] = df[\"Extended Spacer(NNNN[20nt]NGGNNN)\"].apply(extracct_pam)\n",
        "print(\"Unique PAMs:\", df[\"PAM\"].unique())\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "70rjjKfdsX_x",
        "outputId": "8371db6c-279f-41ea-c5a5-8b4bd727083f"
      },
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Unique PAMs: ['GGG' 'CGG' 'TGG' 'AGG']\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "unique_pams = sorted(df[\"PAM\"].unique())\n",
        "pam2id = {p: i for i, p in enumerate(unique_pams)}\n",
        "print(\"pam2id mapping:\", pam2id)\n",
        "\n",
        "pam_ids = df[\"PAM\"].map(pam2id).values\n",
        "\n",
        "X_pam = np.zeros((len(df), len(unique_pams)), dtype=np.float32)\n",
        "X_pam[np.arange(len(df)), pam_ids] = 1.0\n",
        "\n",
        "print(\"X_pam shape:\", X_pam.shape)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "yCpfFWoDtap9",
        "outputId": "8464694d-75b8-4043-920f-ab30966ecab8"
      },
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "pam2id mapping: {'AGG': 0, 'CGG': 1, 'GGG': 2, 'TGG': 3}\n",
            "X_pam shape: (4328, 4)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Encode Gene Symbol as inteder IDs\n",
        "\n",
        "unique_genes = sorted(df[\"Gene Symbol\"].astype(str).unique())\n",
        "gene2id = {g: i for i, g in enumerate(unique_genes)}\n",
        "print(\"Number of unique genes:\", len(unique_genes))\n",
        "\n",
        "gene_ids = df[\"Gene Symbol\"].astype(str).map(gene2id).astype(np.int64).values\n",
        "print(\"gene_ids shape:\", gene_ids.shape, \"min:\", gene_ids.min(), \"max:\", gene_ids.max())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "XdWaBRwLuOMu",
        "outputId": "609d906e-06fc-49dd-b42a-6fc631ec86c5"
      },
      "execution_count": 18,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Number of unique genes: 15\n",
            "gene_ids shape: (4328,) min: 0 max: 14\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Amino Acid Cut Position (scalar + normalization)\n",
        "\n",
        "cut_raw = df[\"Amino Acid Cut position\"].astype(np.float32).values\n",
        "print(\"Cut position eaw stats: min\", cut_raw.min(), \"max:\", cut_raw.max())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "n5RaYeCRvI-8",
        "outputId": "92b78b3f-ce92-461c-c11f-883e22795566"
      },
      "execution_count": 19,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cut position eaw stats: min 1.0 max: 2826.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "y_raw = df[\"Percent Peptide\"].astype(np.float32).values\n",
        "print(\"Percent Peptide eaw stats: min\", y_raw.min(), \"max:\", y_raw.max())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ahjKq-U9v9A7",
        "outputId": "0aaa60d3-f2bf-48bf-94a7-e78304b3107b"
      },
      "execution_count": 20,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Percent Peptide eaw stats: min 0.05 max: 100.0\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "N = len(df)\n",
        "indices = np.arange(N)\n",
        "\n",
        "#60% train, 40% temp\n",
        "train_idx, temp_idx = train_test_split(\n",
        "    indices,\n",
        "    test_size=0.4,\n",
        "    random_state=42,\n",
        "    shuffle=True,\n",
        ")\n",
        "\n",
        "# Split temp into 20 val, 20 test\n",
        "val_idx, test_idx = train_test_split(\n",
        "    temp_idx,\n",
        "    test_size = 0.5,\n",
        "    random_state=42,\n",
        "    shuffle=True,\n",
        ")\n",
        "\n",
        "print(\"Train size:\", len(train_idx), \"Val size:\", len(val_idx), \"Test size\", len(test_idx))"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DzQLp0RywEbc",
        "outputId": "6d8ed678-c657-4732-df4e-a16a1313b13e"
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Train size: 2596 Val size: 866 Test size 866\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Fit scalers on train only and transform all\n",
        "\n",
        "# Reshape to 2D for sklearn\n",
        "cut_raw_2d = cut_raw.reshape(-1, 1)\n",
        "y_raw_2d = y_raw.reshape(-1, 1)\n",
        "\n",
        "cut_scaler = MinMaxScaler()\n",
        "y_scaler = MinMaxScaler()\n",
        "\n",
        "# Fit only on traning subset\n",
        "cut_scaler.fit(cut_raw_2d[train_idx])\n",
        "y_scaler.fit(y_raw_2d[train_idx])\n",
        "\n",
        "#Transform full arrays\n",
        "cut_scaled_full = cut_scaler.transform(cut_raw_2d).astype(np.float32)\n",
        "y_scaled_full = y_scaler.transform(y_raw_2d).astype(np.float32)\n",
        "\n",
        "print(\"Cut scaled range (full):\", cut_scaled_full.min(), cut_scaled_full.max())\n",
        "print(\"y scaled range (full):\", y_scaled_full.min(), y_scaled_full.max())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "xk-Bl5riw0pp",
        "outputId": "a4fbf61f-70dc-45a8-a617-c82634170d91"
      },
      "execution_count": 28,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cut scaled range (full): 0.0 0.99999994\n",
            "y scaled range (full): 0.0 1.0018042\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# slice into train / val / test arrays\n",
        "\n",
        "def split_by_index(arr, train_idx, val_idx, test_idx):\n",
        "  return arr[train_idx], arr[val_idx], arr[test_idx]"
      ],
      "metadata": {
        "id": "QmlyjfDNyXB6"
      },
      "execution_count": 29,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Sequence branch\n",
        "X_seq_train, X_seq_val, X_seq_test = split_by_index(X_seq, train_idx, val_idx, test_idx)\n",
        "\n",
        "#PAM branch (one-hot)\n",
        "X_pam_train, X_pam_val, X_pam_test = split_by_index(X_pam, train_idx, val_idx, test_idx)\n",
        "\n",
        "#Gene IDs (for embedding)\n",
        "X_gene_train, X_gene_val, X_gene_test = split_by_index(gene_ids, train_idx, val_idx, test_idx)\n",
        "\n",
        "# Cut position scalar\n",
        "X_cut_train, X_cut_val, X_cut_test = split_by_index(cut_scaled_full, train_idx, val_idx, test_idx)\n",
        "\n",
        "# Percent Peptide scalar (Targets)\n",
        "y_train, y_val, y_test = split_by_index(y_scaled_full, train_idx, val_idx, test_idx)\n",
        "\n",
        "print(\"X_seq_train shape:\", X_seq_train.shape)\n",
        "print(\"X_pam_train shape:\", X_pam_train.shape)\n",
        "print(\"X_gene_train shape:\", X_gene_train.shape)\n",
        "print(\"X_cut_train shape:\", X_cut_train.shape)\n",
        "print(\"y_train shape:\", y_train.shape)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "CFMWG0DBzVD9",
        "outputId": "221fe4d2-ff73-4930-ab02-9cd5be9d97c1"
      },
      "execution_count": 30,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "X_seq_train shape: (2596, 4, 20)\n",
            "X_pam_train shape: (2596, 4)\n",
            "X_gene_train shape: (2596,)\n",
            "X_cut_train shape: (2596, 1)\n",
            "y_train shape: (2596, 1)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "np.save(\"X_seq_train.npy\", X_seq_train.astype(np.float32))\n",
        "np.save(\"X_seq_val.npy\",   X_seq_val.astype(np.float32))\n",
        "np.save(\"X_seq_test.npy\",  X_seq_test.astype(np.float32))\n",
        "\n",
        "np.save(\"X_pam_train.npy\", X_pam_train.astype(np.float32))\n",
        "np.save(\"X_pam_val.npy\",   X_pam_val.astype(np.float32))\n",
        "np.save(\"X_pam_test.npy\",  X_pam_test.astype(np.float32))\n",
        "\n",
        "np.save(\"X_gene_train.npy\", X_gene_train.astype(np.int64))\n",
        "np.save(\"X_gene_val.npy\",   X_gene_val.astype(np.int64))\n",
        "np.save(\"X_gene_test.npy\",  X_gene_test.astype(np.int64))\n",
        "\n",
        "np.save(\"X_cut_train.npy\", X_cut_train.astype(np.float32))\n",
        "np.save(\"X_cut_val.npy\",   X_cut_val.astype(np.float32))\n",
        "np.save(\"X_cut_test.npy\",  X_cut_test.astype(np.float32))\n",
        "\n",
        "np.save(\"y_train.npy\", y_train.astype(np.float32))\n",
        "np.save(\"y_val.npy\",   y_val.astype(np.float32))\n",
        "np.save(\"y_test.npy\",  y_test.astype(np.float32))"
      ],
      "metadata": {
        "id": "WH4H5RVK2Jdi"
      },
      "execution_count": 31,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "with open(\"gene2id.json\", \"w\") as f:\n",
        "  json.dump(gene2id, f)\n",
        "\n",
        "with open(\"pam2id.json\", \"w\") as f:\n",
        "  json.dump(pam2id, f)\n",
        "\n",
        "# save scaler parameters (so i can inverse-transform later)\n",
        "np.save(\"cut_scaler_min.npy\", cut_scaler.data_min_)\n",
        "np.save(\"cut_scaler_max.npy\", cut_scaler.data_max_)\n",
        "np.save(\"y_scaler_min.npy\",   y_scaler.data_min_)\n",
        "np.save(\"y_scaler_max.npy\",   y_scaler.data_max_)"
      ],
      "metadata": {
        "id": "6ToEf-ft2VKo"
      },
      "execution_count": 32,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [],
      "metadata": {
        "id": "5x0Y-2sv30aV"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}